android/src/main/kotlin/com/example/pocketape/ARCoreManager.kt

```diff
@@ -0,0 +1,92 @@
+import android.content.pm.PackageManager
+import android.util.Log
+import android.widget.Toast
+import androidx.core.content.ContextCompat
+import com.google.ar.core.*
+import com.google.ar.core.exceptions.*
+import com.google.ar.sceneform.ArSceneView
+import com.google.ar.sceneform.Scene
+
+class ARCoreManager(private val context: Context,  private val plugin: PocketapePlugin) {
+    private var session: Session? = null
+    private var arSceneView: ArSceneView? = null
+    private var updateListener: Scene.OnUpdateListener? = null
+
+    private fun initializeARSession() {
+        try {
+            session = Session(context)
+            val config = Config(session).apply {
+                updateMode =  Config.UpdateMode.LATEST_CAMERA_IMAGE
+                depthMode = Config.DepthMode.AUTOMATIC
+            }
+            session?.configure(config)
+        } catch (e: UnavailableArcoreNotInstalledException) {
+            e.printStackTrace()
+        } catch (e: UnavailableApkTooOldException) {
+            e.printStackTrace()
+        } catch (e: UnavailableSdkTooOldException) {
+            e.printStackTrace()
+        } catch (e: UnavailableDeviceNotCompatibleException) {
+            e.printStackTrace()
+        }
+    }
+
+    private fun initializeARSceneView() {
+        arSceneView = ArSceneView(context)
+        arSceneView?.setupSession(session)
+        arSceneView?.scene?.let { scene ->
+            updateListener = Scene.OnUpdateListener {
+                arSceneView?.arFrame?.let { frame -> handleFrameUpdate(frame) }
+            }
+            updateListener?.let { scene.addOnUpdateListener(it) }
+        }
+    }
+
+    private fun handleFrameUpdate(frame: Frame) {
+        sendMeasure(frame)
+    }
+
+    fun startSession() {
+        Log.d("MyTag", "Start")
+        try {
+            if (ContextCompat.checkSelfPermission(context, android.Manifest.permission.CAMERA)
+                == PackageManager.PERMISSION_GRANTED) {
+                initializeARSession()
+                initializeARSceneView()
+                session?.resume()
+                arSceneView?.resume()
+            }
+            else {
+                Toast.makeText(context, "Camera permission is required for AR functionality.", Toast.LENGTH_LONG).show()
+            }
+        } catch (e: Exception) {
+            e.printStackTrace()
+        }
+    }
+
+    fun stopSession() {
+        Log.d("MyTag", "Stop")
+        session?.pause()
+        arSceneView?.pause()
+        onDestroy()
+    }
+
+    private fun sendMeasure(frame: Frame) {
+        var x = frame.camera.displayOrientedPose.translation
+        Log.d("MyTag", "${x[0]} ${x[1]} ${x[2]}")
+        plugin.sendEventToFlutter(x)
+    }
+
+    public fun onDestroy() {
+        arSceneView?.scene?.let { scene ->
+            updateListener?.let { scene.removeOnUpdateListener(it) }
+        }
+        arSceneView?.destroy()
+        session?.close()
+        session = null
+    }
```

Comment:
@Robbendebiene Why do we not pause? It is supposed to pause before removing and destroying to guarantee a proper behavior

Reply:
Any docs you can share that point this out? I would assume that detroying/closing does the pause internally.

Reply:
The library documentation mentioned : Releases resources (a significant amount of native heap memory) used by an ARCore session.
Failure to close sessions explicitly may cause your app to run out of native memory and crash. If your app contains a single AR-enabled activity, it is recommended that you call close() from the activity's onDestroy method.

This method will take several seconds to complete. To prevent blocking the main thread, call pause() on the main thread, and then call close() on a background thread.

It is not safe to call methods on this session or other objects obtained from this session while the session is being closed and after the session is closed.
When closing a session that was created with Session. Feature#SHARED_CAMERA, the camera device must be closed via CameraDevice#close() before calling this method.
Creation of a new session on a different thread will not be blocked by the ongoing call to close() on a background thread.

Reply:
So this is interesting and all, but it doesn't mention that we have to call pause before destroy, or am I missing something? They mention something about performance, but we are anyway executing everything in the same thread so this doesn't apply here.

Reply:
The ArScene should be pause if not the app crush.